sql-plugin/src/main/scala/com/nvidia/spark/rapids/BatchWithPartitionData.scala

```diff
@@ -0,0 +1,356 @@
+        partitionSchema)
+      splitAndCombineBatchWithPartitionData(batch, partitionedGroups, partitionSchema)
+    } else {
+      new SingleGpuColumnarBatchIterator(batch)
+    }
+  }
+
+  /**
+   * Adds a single set of partition values to all rows in a ColumnarBatch.
+   * @return a new columnar batch iterator with partition values
+   */
+  def addSinglePartitionValueToBatch(
+      batch: ColumnarBatch,
+      partitionValues: InternalRow,
+      partitionSchema: StructType): GpuColumnarBatchIterator = {
+    addPartitionValuesToBatch(batch, Array(batch.numRows), Array(partitionValues), partitionSchema)
+  }
+
+  /**
+   * Splits partitions into smaller batches, ensuring that the batch size
+   * for each column does not exceed the specified cuDF limit.
+   *
+   * Data structures:
+   *  - sizeOfBatch: HashMap that stores the size of batches for each column.
+   *  - currentBatch:  Array used to hold the rows and partition values of the current batch.
+   *  - resultBatches: Array that stores the resulting batches after splitting.
+   *
+   * Algorithm:
+   *  - Initialize `sizeOfBatch` and `resultBatches`.
+   *  - Iterate through `partRows`:
+   *    - Add (rows, [values]) to the current batch.
+   *    - For each column:
+   *      - Calculate the size of the current partition for the column.
+   *      - Check if adding this partition to the current batch would exceed cuDF limit.
+   *        - If it does, split the partition:
+   *          - Update the current batch's last value = (new rows, [values]).
+   *          - Add it to the result batches.
+   *          - Process the remaining rows again.
+   *        - Otherwise, increase the `sizeOfBatch` for the column.
+   *
+   * Example:
+   * {{{
+   * Input:
+   *    partition rows:   [10, 40, 70, 10, 11]
+   *    partition values: [ [abc, ab], [bc, ab], [abc, bc], [aa, cc], [ade, fd] ]
+   *    cuDF limit:  300 bytes
+   *
+   * Result:
+   * [
+   *  [ (10, [abc, ab]), (40, [bc, ab]), (63, [abc, bc]) ],
+   *  [ (7, [abc, bc]), (10, [aa, cc]), (11, [ade, fd]) ]
+   * ]
+   * }}}
+   *
+   * @return An array of batches, containing (row counts, partition values) pairs,
+   *         such that each batch's size is less than cuDF limit.
+   */
+  private def splitPartitionDataIntoGroups(partRows: Array[Long], partValues: Array[InternalRow],
+      partSchema: StructType): Array[Array[PartitionRowData]] = {
+    val resultBatches = ArrayBuffer[Array[PartitionRowData]]()
+    val currentBatch = ArrayBuffer[PartitionRowData]()
+    // Initialize a HashMap to store the size of the current batch for each column
+    val sizeOfBatch = mutable.HashMap.empty[Int, Long]
+    var partIndex = 0
+    while (partIndex < partRows.length) {
+      val rowsInPartition = partRows(partIndex)
+      val valuesInPartition = partValues(partIndex)
+      // Add the partition's data to the current batch
+      currentBatch.append(PartitionRowData(valuesInPartition, rowsInPartition))
```

Comment:
Nit: I find this logic a bit more difficult to follow since it adds batches before they're done and then has to "undo" that action.  I think it would be more straightforward if batches were not added until they are fully computed (which already is occurring for `resultBatches`).